File: examples/android/app/src/test/java/org/aetherfeed/app/readerimport/ReaderImportTest.kt
package org.aetherfeed.app.readerimport

import kotlinx.coroutines.runBlocking
import org.aetherfeed.app.data.InMemoryLibrary
import org.aetherfeed.app.domain.Feed
import org.aetherfeed.app.domain.ModuleKind
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class ReaderImportTest {
    private val opml =
        """<opml><body><outline title="Local" xmlUrl="https://example.invalid/rss.xml" htmlUrl="https://example.invalid"/></body></opml>"""

    @Test
    fun detectsTakeoutAndParsesSubscriptions() {
        val file = ReaderImportFile(
            name = "takeout.zip",
            members = mapOf("Takeout/Reader/subscriptions.xml" to opml),
        )
        assertEquals(ReaderVendor.GoogleReader, detectReaderVendor(file))
        val parsed = parseReaderImport(file)
        assertEquals(1, parsed.outlines.size)
        assertEquals("https://example.invalid/rss.xml", parsed.outlines[0].xmlUrl)
    }

    @Test
    fun detectsInoreaderAndGenericOpml() {
        assertEquals(
            ReaderVendor.Inoreader,
            detectReaderVendor(ReaderImportFile("export.opml", "<opml><head><title>Inoreader</title></head>$opml")),
        )
        val parsed = parseReaderImport(ReaderImportFile("feedly.opml", opml))
        assertEquals(ReaderVendor.Opml, parsed.vendor)
        assertEquals(1, parsed.outlines.size)
    }

    @Test
    fun rejectsEmptyAndMalformedWithoutFeeds() {
        val empty = parseReaderImport(ReaderImportFile("empty.opml", "  "))
        assertEquals(ReaderImportErrorCode.Empty, empty.errors[0].code)
        val bad = parseReaderImport(ReaderImportFile("backup.json", "{not-json"))
        assertTrue(bad.errors.any { it.code == ReaderImportErrorCode.Malformed })
        assertTrue(bad.outlines.isEmpty())
    }

    @Test
    fun applySkipsDuplicatesAndLeavesVaultOnHardFailure() = runBlocking {
        val library = InMemoryLibrary()
        library.upsertFeed(
            Feed("existing", "Old", "https://example.invalid/rss.xml", ModuleKind.News, updatedAt = 1),
        )
        val skip = ReaderImportRepository(library).apply(parseReaderImport(ReaderImportFile("feedly.opml", opml)), 10)
        assertEquals(0, skip.feedsAdded)
        assertEquals(1, skip.feedsSkipped)
        val failed = ReaderImportRepository(InMemoryLibrary()).apply(
            parseReaderImport(ReaderImportFile("bad.json", "{nope")),
            10,
        )
        assertEquals(0, failed.feedsAdded)
        assertTrue(failed.errors.any { it.code == ReaderImportErrorCode.Malformed })
    }

    @Test
    fun normalizesDuplicatesAndSortsPodcasts() = runBlocking {
        val library = InMemoryLibrary()
        val mixed = """
            <opml><body>
            <outline text="Podcasts"><outline title="Cast" xmlUrl="https://feeds.libsyn.com/9/rss"/></outline>
            <outline title="News" xmlUrl="HTTP://WWW.example.invalid/rss.xml/"/>
            <outline title="Dup" xmlUrl="https://example.invalid/rss.xml"/>
            </body></opml>
        """.trimIndent()
        val result = ReaderImportRepository(library).apply(parseReaderImport(ReaderImportFile("greader.opml", mixed)), 10)
        assertEquals(2, result.feedsAdded)
        assertEquals(1, result.feedsSkipped)
        assertEquals(1, result.newsAdded)
        assertEquals(1, result.podcastsAdded)
        assertEquals("https://example.invalid/rss.xml", normalizeFeedUrl("HTTP://WWW.example.invalid/rss.xml/"))
        val zip = bytesToImportFile("empty.zip", ByteArray(0))
        assertEquals("", zip.text)
    }

    @Test
    fun parsesInoreaderJsonAndTakeoutStars() {
        val ino = parseReaderImport(
            ReaderImportFile(
                "inoreader-backup.json",
                """{"feeds":[{"title":"Local","feedUrl":"https://example.invalid/rss.xml"}]}""",
            ),
        )
        assertEquals(ReaderVendor.Inoreader, ino.vendor)
        assertEquals("https://example.invalid/rss.xml", ino.outlines[0].xmlUrl)
        val members = mapOf(
            "Takeout/Reader/subscriptions.xml" to opml,
            "Takeout/Reader/starred.json" to
                """{"title":"Starred items","items":[{"title":"Saved","canonical":[{"href":"https://example.invalid/s"}]}]}""",
        )
        assertEquals(opml, takeoutSubscriptionsXml(members))
        val takeout = parseReaderImport(ReaderImportFile("takeout.zip", members = members))
        assertEquals(1, takeout.stars.size)
        assertTrue(readerImportSummary(ReaderImportResult(ReaderVendor.GoogleReader, 1, 0, 1, emptyList())).contains("stars=1"))
    }

    @Test
    fun parsesAetherFeedLibraryKindAndFolder() = runBlocking {
        val parsed = parseReaderImport(
            ReaderImportFile(
                "seed-library.json",
                """[{"title":"Local","url":"https://example.invalid/rss.xml","kind":"news","folder":"Art"}]""",
            ),
        )
        assertEquals("Art", parsed.outlines.single().folder)
        assertEquals("news", parsed.outlines.single().type)
        val planned = planReaderImport(emptySet(), parsed, 1)
        assertEquals("Art", planned.feeds.single().folder)
        assertEquals(ModuleKind.News, planned.feeds.single().kind)
    }
}

File: examples/web/src/readerimport/readerImport.test.ts
import {
  classifyFeedKind,
  createReaderImportRepository,
  detectReaderVendor,
  type Feed,
  normalizeFeedUrl,
  parseReaderImport,
  type ReadState,
  type Star,
} from "@aetherfeed/domain";
import { describe, expect, it } from "vitest";

const OPML = `<opml><body><outline title="Local" xmlUrl="https://example.invalid/rss.xml"/></body></opml>`;

describe("reader import API", () => {
  it("detects vendors and parses OPML", () => {
    expect(
      detectReaderVendor({ name: "takeout.zip", members: { "subscriptions.xml": OPML } }),
    ).toBe("google-reader");
    expect(detectReaderVendor({ name: "inoreader-backup.json", text: '{"feeds":[]}' })).toBe(
      "inoreader",
    );
    const parsed = parseReaderImport({ name: "feedly.opml", text: OPML });
    expect(parsed.vendor).toBe("opml");
    expect(parsed.outlines[0]?.xmlUrl).toBe("https://example.invalid/rss.xml");
  });

  it("rejects empty files and skips duplicate URLs", async () => {
    expect(parseReaderImport({ name: "empty.opml", text: "" }).errors[0]?.code).toBe("empty");
    const feeds: Feed[] = [
      { id: "x", title: "Old", url: "https://example.invalid/rss.xml", kind: "news", updatedAt: 1 },
    ];
    const library = {
      feeds: async () => feeds,
      upsertFeed: async (feed: Feed) => {
        feeds.push(feed);
      },
      upsertStar: async (_star: Star) => undefined,
      upsertReadState: async (_state: ReadState) => undefined,
    };
    const result = await createReaderImportRepository(library).apply(
      parseReaderImport({ name: "feedly.opml", text: OPML }),
      10,
    );
    expect(result.feedsSkipped).toBe(1);
    expect(result.feedsAdded).toBe(0);
  });

  it("normalizes URLs and sorts podcast folders", async () => {
    expect(normalizeFeedUrl("HTTP://WWW.Example.invalid/rss.xml/")).toBe(
      normalizeFeedUrl("https://example.invalid/rss.xml"),
    );
    expect(normalizeFeedUrl("feed/https://www.example.invalid:443/rss.xml")).toBe(
      "https://example.invalid/rss.xml",
    );
    expect(
      classifyFeedKind({
        title: "Cast",
        xmlUrl: "https://feeds.libsyn.com/9/rss",
        folder: "Podcasts",
      }),
    ).toBe("podcast");
    const feeds: Feed[] = [];
    const library = {
      feeds: async () => feeds,
      upsertFeed: async (feed: Feed) => {
        feeds.push(feed);
      },
      upsertStar: async (_star: Star) => undefined,
      upsertReadState: async (_state: ReadState) => undefined,
    };
    const result = await createReaderImportRepository(library).apply(
      parseReaderImport({
        name: "inoreader.opml",
        text: `<opml><body>
          <outline text="Podcasts"><outline title="Cast" xmlUrl="https://feeds.libsyn.com/9/rss"/></outline>
          <outline title="A" xmlUrl="https://www.example.invalid/rss.xml/"/>
          <outline title="B" xmlUrl="http://example.invalid/rss.xml"/>
        </body></opml>`,
      }),
      10,
    );
    expect(result.feedsAdded).toBe(2);
    expect(result.feedsSkipped).toBe(1);
    expect(result.podcastsAdded).toBe(1);
    expect(result.newsAdded).toBe(1);
    expect(feeds.find((feed) => feed.kind === "podcast")?.folder).toBe("Podcasts");
  });
});
